p2p/address-book/src/peer_list.rs

```diff
@@ -78,37 +79,36 @@ impl<Z: NetworkZone> PeerList<Z> {
         }
     }
 
-    /// Gets a reference to a peer
-    pub fn get_peer(&self, peer: &Z::Addr) -> Option<&ZoneSpecificPeerListEntryBase<Z::Addr>> {
-        self.peers.get(peer)
-    }
-
     /// Returns a random peer.
     /// If the pruning seed is specified then we will get a random peer with
     /// that pruning seed otherwise we will just get a random peer in the whole
     /// list.
-    pub fn get_random_peer<R: Rng>(
-        &self,
+    ///
+    /// The given peer will be removed from the peer list.
+    pub fn take_random_peer<R: Rng>(
+        &mut self,
         r: &mut R,
         block_needed: Option<u64>,
-    ) -> Option<&ZoneSpecificPeerListEntryBase<Z::Addr>> {
+    ) -> Option<ZoneSpecificPeerListEntryBase<Z::Addr>> {
         if let Some(needed_height) = block_needed {
             let (_, addresses_with_block) = self.pruning_seeds.iter().find(|(seed, _)| {
                 // TODO: factor in peer blockchain height?
                 seed.get_next_unpruned_block(needed_height, CRYPTONOTE_MAX_BLOCK_HEIGHT)
-                    .expect("Explain")
+                    .expect("Block needed is higher than max block allowed.")
                     == needed_height
             })?;
             let n = r.gen_range(0..addresses_with_block.len());
-            self.get_peer(&addresses_with_block[n])
+            let peer = addresses_with_block[n];
+            self.remove_peer(&peer)
         } else {
             let len = self.len();
             if len == 0 {
                 None
             } else {
                 let n = r.gen_range(0..len);
 
-                self.peers.values().nth(n)
+                let (&key, _) = self.peers.get_index(n).unwrap();
```

Comment:
Does this work?
```suggestion
                let (&key, _) = self.peers[n];
```

Reply:
`[n]` just returns the value, although we could get the key from the value it would require copying the whole value.